protos/icmp/icmp.go

```diff
@@ -0,0 +1,329 @@
+	} else {
+		icmp.publishTransaction(trans)
+	}
+}
+
+func (icmp *Icmp) processResponse(tuple *IcmpTuple, msg *IcmpMessage) {
+	logp.Debug("icmp", "Processing response. %s", tuple)
+
+	revTuple := tuple.Reverse()
+	trans := icmp.deleteTransaction(revTuple.Hashable())
+	if trans == nil {
+		trans = newTransaction(msg.Ts, revTuple)
+		trans.Notes = append(trans.Notes, OrphanedResponseMsg)
+		logp.Debug("icmp", OrphanedResponseMsg+" %s", tuple)
+	}
+
+	trans.Response = msg
+	icmp.publishTransaction(trans)
+}
+
+func (icmp *Icmp) direction(t *IcmpTransaction) uint8 {
+	if !icmp.isLocalIp(t.Tuple.SrcIp) {
+		return IcmpDirectionFromOutside
+	}
+	if !icmp.isLocalIp(t.Tuple.DstIp) {
+		return IcmpDirectionFromInside
+	}
+	return IcmpDirectionLocalOnly
+}
+
+func (icmp *Icmp) isLocalIp(ip net.IP) bool {
+	if ip.IsLoopback() {
+		return true
+	}
+
+	for _, localIp := range icmp.localIps {
+		if ip.Equal(localIp) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (icmp *Icmp) getTransaction(k HashableIcmpTuple) *IcmpTransaction {
+	v := icmp.transactions.Get(k)
+	if v != nil {
+		return v.(*IcmpTransaction)
+	}
+	return nil
+}
+
+func (icmp *Icmp) deleteTransaction(k HashableIcmpTuple) *IcmpTransaction {
+	v := icmp.transactions.Delete(k)
+	if v != nil {
+		return v.(*IcmpTransaction)
+	}
+	return nil
+}
+
+func (icmp *Icmp) expireTransaction(tuple HashableIcmpTuple, trans *IcmpTransaction) {
+	trans.Notes = append(trans.Notes, OrphanedRequestMsg)
+	logp.Debug("icmp", OrphanedRequestMsg+" %s", &trans.Tuple)
+	icmp.publishTransaction(trans)
+}
+
+func (icmp *Icmp) publishTransaction(trans *IcmpTransaction) {
+	if icmp.results == nil {
+		return
+	}
+
+	logp.Debug("icmp", "Publishing transaction. %s", &trans.Tuple)
+
+	event := common.MapStr{}
+
+	// common fields - group "env"
+	event["client_ip"] = trans.Tuple.SrcIp
+	event["ip"] = trans.Tuple.DstIp
+
+	// common fields - group "event"
+	event["timestamp"] = common.Time(trans.Ts) // timestamp of the first packet
+	event["type"] = "icmp"                     // protocol name
+	event["count"] = 1                         // reserved for future sampling support
+	event["path"] = trans.Tuple.DstIp          // what is requested (dst ip)
+	if trans.HasError() {
+		event["status"] = common.ERROR_STATUS
+	} else {
+		event["status"] = common.OK_STATUS
+	}
+	if len(trans.Notes) > 0 {
+		event["notes"] = trans.Notes
+	}
+
+	// common fields - group "measurements"
+	responsetime, hasResponseTime := trans.ResponseTimeMillis()
+	if hasResponseTime {
+		event["responsetime"] = responsetime
+	}
+	switch icmp.direction(trans) {
+	case IcmpDirectionFromInside:
+		if trans.Request != nil {
+			event["bytes_out"] = trans.Request.Length
+		}
+		if trans.Response != nil {
+			event["bytes_in"] = trans.Response.Length
+		}
+	case IcmpDirectionFromOutside:
+		if trans.Request != nil {
+			event["bytes_in"] = trans.Request.Length
+		}
+		if trans.Response != nil {
+			event["bytes_out"] = trans.Response.Length
+		}
```

Comment:
switch-case incomplete. IcmpDirectionLocalOnly not checked. "bytes_in" and "bytes_out" missing?